jupyterhub/services/auth.py

```diff
@@ -351,7 +364,47 @@ def _default_scopes(self):
             return {f'access:services!service={service_name}'}
         return set()
 
-    def _check_hub_authorization(self, url, api_token, cache_key=None, use_cache=True):
+    _pool = Any(help="Thread pool for running async methods in the background")
+
+    @default("_pool")
+    def _new_pool(self):
+        # start a single ThreadPool in the background
+        from concurrent.futures import ThreadPoolExecutor
+
+        pool = ThreadPoolExecutor(1)
+        # create an event loop in the thread
+        pool.submit(self._setup_asyncio_thread).result()
+        return pool
+
+    def _setup_asyncio_thread(self):
+        """Create asyncio loop
+
+        To be called from the background thread,
+        so that any thread-local state is setup correctly
+        """
+        self._thread_loop = asyncio.new_event_loop()
+
+    def _synchronize(self, async_f, *args, **kwargs):
+        """Call an async method in our background thread"""
+        future = self._pool.submit(
+            lambda: self._thread_loop.run_until_complete(async_f(*args, **kwargs))
+        )
+        return future.result()
+
+    def _maybe_async(self, sync, f, *args, **kwargs):
+        """Call an async function, either blocking or returning an awaitable
+
+        if sync: calls function directly, returning awaitable
+        else: Block on a call in our background thread, return actual result
+        """
```

Comment:
`maybe_async` doesn't occur elsewhere in jupyterhub. Maybe in another package?

WDYT about `_maybe_await` because it _might_ await, or it might just return the called coroutine? Or `_call_coroutine(sync, coro_f, ...)` Because this is how we call our coroutine methods, which may or may not be done in a synchronous way?

Reply:
Now using `_call_coroutine`

Reply:
Thanks for thinking about this, sounds good to me!